File: sources/KyinoSense 17.1/net/ccbluex/liquidbounce/features/module/modules/render/BlockESP.java
/*
 * Decompiled with CFR 0.152.
 * 
 * Could not load the following classes:
 *  net.minecraft.block.Block
 *  net.minecraft.init.Blocks
 *  net.minecraft.util.BlockPos
 */
package net.ccbluex.liquidbounce.features.module.modules.render;

import java.awt.Color;
import java.util.ArrayList;
import java.util.List;
import net.ccbluex.liquidbounce.event.EventTarget;
import net.ccbluex.liquidbounce.event.Render3DEvent;
import net.ccbluex.liquidbounce.event.UpdateEvent;
import net.ccbluex.liquidbounce.features.module.Module;
import net.ccbluex.liquidbounce.features.module.ModuleCategory;
import net.ccbluex.liquidbounce.features.module.ModuleInfo;
import net.ccbluex.liquidbounce.utils.block.BlockUtils;
import net.ccbluex.liquidbounce.utils.render.ColorUtils;
import net.ccbluex.liquidbounce.utils.render.RenderUtils;
import net.ccbluex.liquidbounce.utils.timer.MSTimer;
import net.ccbluex.liquidbounce.value.BlockValue;
import net.ccbluex.liquidbounce.value.BoolValue;
import net.ccbluex.liquidbounce.value.IntegerValue;
import net.ccbluex.liquidbounce.value.ListValue;
import net.minecraft.block.Block;
import net.minecraft.init.Blocks;
import net.minecraft.util.BlockPos;

@ModuleInfo(name="BlockESP", description="Allows you to see a selected block through walls.", category=ModuleCategory.RENDER)
public class BlockESP
extends Module {
    private final ListValue modeValue = new ListValue("Mode", new String[]{"Box", "2D"}, "Box");
    private final BlockValue blockValue = new BlockValue("Block", 168);
    private final IntegerValue radiusValue = new IntegerValue("Radius", 40, 5, 120);
    private final IntegerValue colorRedValue = new IntegerValue("R", 255, 0, 255);
    private final IntegerValue colorGreenValue = new IntegerValue("G", 179, 0, 255);
    private final IntegerValue colorBlueValue = new IntegerValue("B", 72, 0, 255);
    private final BoolValue colorRainbow = new BoolValue("Rainbow", false);
    private final MSTimer searchTimer = new MSTimer();
    private final List<BlockPos> posList = new ArrayList<BlockPos>();
    private Thread thread;

    @EventTarget
    public void onUpdate(UpdateEvent event) {
        if (this.searchTimer.hasTimePassed(1000L) && (this.thread == null || !this.thread.isAlive())) {
            int radius = (Integer)this.radiusValue.get();
            Block selectedBlock = Block.func_149729_e((int)((Integer)this.blockValue.get()));
            if (selectedBlock == null || selectedBlock == Blocks.field_150350_a) {
                return;
            }
            this.thread = new Thread(() -> {
                ArrayList<BlockPos> blockList = new ArrayList<BlockPos>();
                for (int x = -radius; x < radius; ++x) {
                    for (int y = radius; y > -radius; --y) {
                        for (int z = -radius; z < radius; ++z) {
                            int xPos = (int)BlockESP.mc.field_71439_g.field_70165_t + x;
                            int yPos = (int)BlockESP.mc.field_71439_g.field_70163_u + y;
                            int zPos = (int)BlockESP.mc.field_71439_g.field_70161_v + z;
                            BlockPos blockPos = new BlockPos(xPos, yPos, zPos);
                            Block block = BlockUtils.getBlock(blockPos);
                            if (block != selectedBlock) continue;
                            blockList.add(blockPos);
                        }
                    }
                }
                this.searchTimer.reset();
                List<BlockPos> list = this.posList;
                synchronized (list) {
                    this.posList.clear();
                    this.posList.addAll(blockList);
                }
            }, "BlockESP-BlockFinder");
            this.thread.start();
        }
    }

    /*
     * WARNING - Removed try catching itself - possible behaviour change.
     */
    @EventTarget
    public void onRender3D(Render3DEvent event) {
        List<BlockPos> list = this.posList;
        synchronized (list) {
            Color color = (Boolean)this.colorRainbow.get() != false ? ColorUtils.rainbow() : new Color((Integer)this.colorRedValue.get(), (Integer)this.colorGreenValue.get(), (Integer)this.colorBlueValue.get());
            for (BlockPos blockPos : this.posList) {
                switch (((String)this.modeValue.get()).toLowerCase()) {
                    case "box": {
                        RenderUtils.drawBlockBox(blockPos, color, true);
                        break;
                    }
                    case "2d": {
                        RenderUtils.draw2D(blockPos, color.getRGB(), Color.BLACK.getRGB());
                    }
                }
            }
        }
    }

    @Override
    public String getTag() {
        return BlockUtils.getBlockName((Integer)this.blockValue.get());
    }
}

File: sources/Kevin Reborn/kevin/module/modules/render/BlockESP.kt
/*
 * This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package kevin.module.modules.render

import kevin.event.EventTarget
import kevin.event.Render3DEvent
import kevin.event.UpdateEvent
import kevin.main.KevinClient
import kevin.module.*
import kevin.module.modules.misc.Teams
import kevin.utils.BlockUtils.getBlock
import kevin.utils.BlockUtils.getBlockName
import kevin.utils.ColorUtils.rainbow
import kevin.utils.MSTimer
import kevin.utils.RenderUtils
import net.minecraft.block.Block
import net.minecraft.init.Blocks
import net.minecraft.util.BlockPos
import java.awt.Color
import java.util.*

class BlockESP : Module("BlockESP", "Allows you to see a selected block through walls.", category = ModuleCategory.RENDER) {
    private val modeValue = ListValue("Mode", arrayOf("Box", "2D"), "Box")
    private val blockValue = BlockValue("Block", 168)
    private val radiusValue = IntegerValue("Radius", 40, 5, 120)
    private val blockLimitValue = IntegerValue("BlockLimit", 256, 0, 2056)
    private val colorRedValue = IntegerValue("R", 255, 0, 255)
    private val colorGreenValue = IntegerValue("G", 179, 0, 255)
    private val colorBlueValue = IntegerValue("B", 72, 0, 255)
    private val colorRainbow = BooleanValue("Rainbow", false)
    private val searchTimer = MSTimer()
    private val posList: MutableList<BlockPos> = ArrayList()
    private var thread: Thread? = null

    @EventTarget
    fun onUpdate(event: UpdateEvent?) {
        if (searchTimer.hasTimePassed(1000L) && (thread == null || !thread!!.isAlive)) {
            val radius = radiusValue.get()
            val selectedBlock = Block.getBlockById(blockValue.get())

            if (selectedBlock == null || selectedBlock == Blocks.air)
                return

            thread = Thread({
                val blockList: MutableList<BlockPos> = ArrayList()

                for (x in -radius until radius) {
                    for (y in radius downTo -radius + 1) {
                        for (z in -radius until radius) {
                            val thePlayer = mc.thePlayer!!

                            val xPos = thePlayer.posX.toInt() + x
                            val yPos = thePlayer.posY.toInt() + y
                            val zPos = thePlayer.posZ.toInt() + z

                            val blockPos = BlockPos(xPos, yPos, zPos)
                            val block = getBlock(blockPos)

                            if (block == selectedBlock && blockList.size < blockLimitValue.get()) blockList.add(blockPos)
                        }
                    }
                }
                searchTimer.reset()

                synchronized(posList) {
                    posList.clear()
                    posList.addAll(blockList)
                }
            }, "BlockESP-BlockFinder")

            thread!!.start()
        }
    }

    @EventTarget
    fun onRender3D(event: Render3DEvent?) {
        synchronized(posList) {
            val teams = KevinClient.moduleManager.getModule(Teams::class.java)
            for (blockPos in posList) {
                val color =
                    if (teams.bedCheckValue.get() && blockPos in teams.teamBed) Color.green else if (colorRainbow.get()) rainbow() else Color(
                        colorRedValue.get(),
                        colorGreenValue.get(),
                        colorBlueValue.get()
                    )
                when (modeValue.get().lowercase(Locale.getDefault())) {
                    "box" -> RenderUtils.drawBlockBox(blockPos, color, true)
                    "2d" -> RenderUtils.draw2D(blockPos, color.rgb, Color.BLACK.rgb)
                }
            }
        }
    }

    override val tag: String
        get() = getBlockName(blockValue.get())
}